# mnist Data Classification using vaiours ML models

### Import and load the data and the required packages

*Here we are importing the mnist data and calculating the accuracy of the variious ML models*

In [1]:
from sklearn import svm
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import accuracy_score,confusion_matrix

df_train=pd.read_csv('mnist_train.csv')
df_test=pd.read_csv('mnist_test.csv')

X_train=df_train[df_train.columns[1:]]
X_test=df_test[df_test.columns[1:]]
X_test.shape

(10000, 784)

In [2]:
y_train=df_train['label']
y_test=df_test['label']

In [3]:
scalar=MinMaxScaler(feature_range=(-1,1)).fit(X_train)
scalar
X_train=scalar.transform(X_train)
X_test=scalar.transform(X_test)

# SVM

In [7]:
rbf_svm=svm.SVC(kernel='rbf')

In [8]:
rbf_svm = svm.SVC(kernel='rbf')
rbf_svm.fit(X_train, y_train)

SVC()

In [9]:
y_pred=rbf_svm.predict(X_test)


In [10]:
y_pred

array([7, 2, 1, ..., 4, 5, 6], dtype=int64)

In [11]:
accuracy_scoreracy_scorecy_testcy_score(y_pred,y_test)
sc

0.9791

# KNN classifer

In [5]:
from sklearn.neighbors import KNeighborsClassifier

clas=KNeighborsClassifier(n_neighbors=10)
clas.fit(X_train,y_train)
y_pred=clas.predict(X_test)
y_pred

array([7, 2, 1, ..., 4, 5, 6], dtype=int64)

In [6]:
acc=accuracy_score(y_test,y_pred)
acc

0.9664

In [8]:
cm=confusion_matrix(y_test,y_pred)
cm

array([[ 971,    1,    1,    0,    0,    2,    4,    1,    0,    0],
       [   0, 1132,    2,    0,    0,    0,    1,    0,    0,    0],
       [  13,   12,  982,    2,    1,    0,    2,   17,    3,    0],
       [   0,    3,    3,  976,    1,   10,    1,    7,    6,    3],
       [   2,   11,    0,    0,  940,    0,    4,    1,    1,   23],
       [   4,    0,    0,   12,    1,  863,    6,    1,    1,    4],
       [   6,    4,    0,    0,    3,    2,  943,    0,    0,    0],
       [   0,   27,    4,    0,    2,    0,    0,  983,    0,   12],
       [   6,    4,    5,   11,    7,    9,    4,    7,  914,    7],
       [   7,    6,    3,    7,   10,    3,    1,   10,    2,  960]],
      dtype=int64)

In [15]:
for i in range(len(y_test)):
    if y_test[i] != y_pred[i]:
        print(y_test[i],'   ',y_pred[i])

4     0
2     1
8     2
2     7
4     9
3     1
9     8
4     6
4     1
2     0
9     7
2     7
6     4
7     4
3     7
2     8
6     0
4     9
2     1
3     7
9     3
8     0
8     1
7     1
4     9
8     3
2     7
3     9
2     1
2     6
2     7
7     1
8     4
4     9
4     9
4     9
5     9
8     3
8     3
2     7
3     5
2     0
8     9
5     4
3     5
6     0
6     5
7     1
3     7
8     4
9     8
4     6
7     9
6     1
2     1
9     4
7     2
4     9
9     0
7     1
7     1
5     9
5     7
8     3
7     1
8     3
5     6
5     3
8     6
8     9
8     1
4     1
7     1
7     9
8     7
9     3
9     3
3     7
9     7
4     9
9     3
7     1
8     0
7     9
5     3
7     1
8     5
8     4
2     7
8     5
8     9
8     7
4     9
8     3
9     4
8     2
8     1
2     0
4     8
2     7
8     4
4     9
7     1
7     9
2     1
8     1
2     0
3     4
6     0
9     2
4     9
6     1
8     2
8     3
1     2
0     5
2     0
9     1
4     1
5     6
9     6
2     7
7     1
5     3
9     1


In [16]:
con=0
for i in range(len(y_test)):
    if y_test[i] == y_pred[i]:
        con+=1
con

9664

In [17]:
(10000-9664)/10000

0.0336

In [18]:
1-0.0336

0.9664

In [25]:
y_test.count


<bound method Series.count of 0       7
1       2
2       1
3       0
4       4
       ..
9995    2
9996    3
9997    4
9998    5
9999    6
Name: label, Length: 10000, dtype: int64>

# Decision Tree Classifier

In [26]:
from sklearn.tree import DecisionTreeClassifier

class_dt=DecisionTreeClassifier(random_state=9)
class_dt.fit(X_train,y_train)
y_pred_df=class_dt.predict(X_test)
confusion_matrix(y_test,y_pred_df)

class_dt.score(X_test,y_test)

0.8756

# Logistic Regression

In [29]:
from sklearn.linear_model import LogisticRegression

classifier=LogisticRegression()
classifier.fit(X_train,y_train)

y_pred=classifier.predict(X_test)
classifier.score(X_test,y_test)

C:\Users\sreek\AppData\Local\Continuum\anaconda3\envs\gpuenv\lib\site-packages\sklearn\linear_model\_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


0.9231

## conclusion

> **From all the above models the SVM gives the better accurancy over 97% for the minist data**